### PR TITLE
cmake: enable building and running catch-style unit test suite. 

### PR DESCRIPTION
**Pull request by dankegel**
After doing 'make', to run the tests, do 'make test' (shows little output) or 'ctest -v' (shows more output) or
Debug/TestMain (bare metal, lets you use catch.hpp options).

### CODE REVIEW

**Comment by lucasvickers** [comment]
- On OS X, the binary get's embedded inside the .app folder `./Debug/TestMain.app/Contents/MacOS/TestMain`, any easy way to add that to the search folders for `TestMain`?

- Works on the linux side except ctest seems to suppress the output when run via `make test`.  This is pretty easily resolved with running via `ctest --verbose`.  Not sure if we want to show output by default?

- Any reason you didn't include the unit tests in the subfolders [audio](https://github.com/cinder/Cinder/tree/master/test/unit/src/audio) and [signals](https://github.com/cinder/Cinder/tree/master/test/unit/src/signals)?


Thanks!  This is going to be really useful for our CI.  Removes the need for an additional testing script.

**Comment by richardeakin** [comment]
Yea, this brings to light a little problem with the assets folder, in that the app is expected to live relative to the assets folder. If you're building out-of-source then the app gets built in `${CINDER_REPO}/build/Debug/...`, so it can't find the `${CINDER_REPO}/test/unit/assets` folder. Need to think on this..

**Comment by lucasvickers** [comment]
Just to clarify, I meant for `ctest` to find the binary, otherwise you see:
```
$ make test
Running tests...
Test project /code/lib/ctest/build
    Start 1: TestMain
Could not find executable /code/lib/ctest/build/TestMain
Looked in the following places:
/code/lib/ctest/build/TestMain
/code/lib/ctest/build/TestMain
/code/lib/ctest/build/Release/TestMain
/code/lib/ctest/build/Release/TestMain
/code/lib/ctest/build/Debug/TestMain
/code/lib/ctest/build/Debug/TestMain
/code/lib/ctest/build/MinSizeRel/TestMain
/code/lib/ctest/build/MinSizeRel/TestMain
/code/lib/ctest/build/RelWithDebInfo/TestMain
/code/lib/ctest/build/RelWithDebInfo/TestMain
/code/lib/ctest/build/Deployment/TestMain
/code/lib/ctest/build/Deployment/TestMain
/code/lib/ctest/build/Development/TestMain
/code/lib/ctest/build/Development/TestMain
code/lib/ctest/build/TestMain
code/lib/ctest/build/TestMain
code/lib/ctest/build/Release/TestMain
code/lib/ctest/build/Release/TestMain
code/lib/ctest/build/Debug/TestMain
code/lib/ctest/build/Debug/TestMain
code/lib/ctest/build/MinSizeRel/TestMain
code/lib/ctest/build/MinSizeRel/TestMain
code/lib/ctest/build/RelWithDebInfo/TestMain
code/lib/ctest/build/RelWithDebInfo/TestMain
code/lib/ctest/build/Deployment/TestMain
code/lib/ctest/build/Deployment/TestMain
code/lib/ctest/build/Development/TestMain
code/lib/ctest/build/Development/TestMain
Unable to find executable: /code/lib/ctest/build/TestMain
1/1 Test #1: TestMain .........................***Not Run   0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.00 sec

The following tests FAILED:
	  1 - TestMain (Not Run)
Errors while running CTest
make: *** [test] Error 8
```

**Comment by dankegel** [comment]
Try it now.  "make check" should give you the verbose output we both crave :-),
it should find the tests on MacOSX, and it adds the tests I missed.

File: test/unit/src/audio/FftUnit.cpp [hunk 1]
```diff
@@ -3,7 +3,7 @@
 // FIXME: OOURA roundtrip FFT seems to be broken on windows for sizeFft = 4 (https://github.com/cinder/Cinder/issues/1263)
 #if ! defined( CINDER_MSW )
 
-#include "catch.hpp"
+#include "../catch.hpp"
```
Comment (sansumbrella) [inline test/unit/src/audio/FftUnit.cpp]: Shouldn't the header search path for the tests be consistent so we don't need to describe relative paths in the test files themselves?

**Comment by sansumbrella** [comment]
Yay unit tests!

File: test/unit/src/audio/FftUnit.cpp (hunk 1 above)
Comment (dankegel) [inline test/unit/src/audio/FftUnit.cpp]: If you want to move catch.hpp somewhere more sensible, I'm fine with that.  Where do you want it?

Or are suggesting adding a -I .. when building the subdirs?  That'd be kinda ugly.

**Comment by lucasvickers** [comment]
The cmake functionality works really well!  Note: there are some unix unit test issues, opening #1551 to track those.  Pending the ongoing comments about header location and asset directories this looks good to me.

**Comment by dankegel** [comment]
Not sure I get the asset directory one.  Are you objecting to sprinkling
app::Platform::get()->addAssetDirectory(UNIT_DIR_ASSETS);
in each file that uses assets?  I'd love a cleaner solution.

Note that the tests pass for me on MacOS X now, so the above is probably sufficient.

File: test/unit/src/audio/FftUnit.cpp (hunk 1 above)
Comment (richardeakin) [inline test/unit/src/audio/FftUnit.cpp]: I'd imagine adding `INCLUDES "${UNIT_DIR}/src"` when calling `ci_make_app()` would do it. I can't really think of a more reasonable place for catch.hpp to live other than `cinder/test/unit/src/catch.hpp` but of course open to other places.

**Comment by lucasvickers** [comment]
@richardeakin are you okay with `app::Platform::get()->addAssetDirectory(UNIT_DIR_ASSETS);` for now?  I know long term we're talking about increasing the number of parent directories an app will search, but for now it's just isolated to the unit tests.

@dankegel basically it just brought back up a convo that with the cmake folder structures we need to support searching further up the folder index.

File: test/unit/src/audio/FftUnit.cpp (hunk 1 above)
Comment (sansumbrella) [inline test/unit/src/audio/FftUnit.cpp]: I'm not familiar enough with CMAKE to say how to set the search paths or for it to be clear what you mean by build the subdirs.

The catch unit tests can all be (and should be, I believe) compiled into a single binary that can select different test cases to run. So I visualize the file hierarchy as a bunch of things relative to the root of the test project. Effectively, you're building one thing from the root of the `src/` directory and linking in things from the subdirectories. So in XCode, there's just one project at the root and all the include directories are handled therer.

**Comment by dankegel** [comment]
fun fact: increasing the number of parent directories wouldn't help here.

**Comment by lucasvickers** [comment]
Whoops, you are correct.

File: test/unit/src/audio/FftUnit.cpp (hunk 1 above)
Comment (dankegel) [inline test/unit/src/audio/FftUnit.cpp]: This patch does indeed build a single catch unit test executable, just as you expected.

Comment (richardeakin) [inline test/unit/src/audio/FftUnit.cpp]: Actually, I can think of a better place to put the catch header - at `cinder/test/unit/include/catch.hpp`. The benefit here that I see is that if other projects that already depend on cinder want to use catch.hpp, they can add that folder to their include path without picking up any of cinder's unit tests.  But the move can be done outside of this PR, and probably should as the xcode and VS project files will need to be updated.

**Comment by dankegel** [comment]
Right then, have a look now.  No more nasty ../check.h's.

**Comment by richardeakin** [comment]
This is looking great. Open to merging now, but wanted to bring up a couple points first:

1)  Should the `enable_testing()` and `add_custom_targets()` in the main CMakeLists.txt file only be added if a certain cache variable is defined, say `CINDER_TEST_ENABLE` or something like that? Seems that it would speed up builds a bit for those just wanting to rebuild libcinder.

2) Fine for now but it seems like making the .app bundle on OS X isn't needed here, since it is meant to be run from the command line. Perhaps in the future we need some sort of `ci_make_app()` variant for command line apps, or give it an extra BOOL argument so it doesn't do all the `MACOSX_BUNDLE WIN32` like things.

**Comment by dankegel** [comment]
how about CINDER_DISABLE_TESTS, on the theory that running tests should be on the easy path?

**Comment by richardeakin** [comment]
Personally, I think the majority of people don't want to build the tests - both people like me who are frequently rebuilding cinder while prototyping various things, or common users who just want to build master branch, shouldn't have to worry about tests. We're planning to have CI do the worrying for us. :)

I was also basing the name and default on what the author of [GLM chose](https://github.com/g-truc/glm/blob/c947b3338bf4442cdc41d70785c7f0d645e750f5/CMakeLists.txt#L29), as I think it works well there.

**Comment by dankegel** [comment]
I added CINDER_BUILD_TESTS next to the existing CINDER_BUILD_SAMPLES, how's that look?

**Comment by richardeakin** [comment]
Yea that makes sense to me, grouping them that way. Looks ready to merge to me!

Thanks again for working on this, really nice to see us moving towards more automation and less manual labor.

**Comment by dankegel** [comment]
I can't wait for it to land, it'll make working on issue 1544 easier.

**Pull request merged**